### ISSUE TITLE
API generification

### ISSUE THREAD

**Issue by vietj** [reactiverse/reactive-pg-client#245]
This issue is rather an open discussion about how the client will handle multiple back-ends, the concerns we want to address are:

* naming the project and the package
* how to handle specific backend operations (PG has notifications/notices, MY has ping, etc...)

## Naming

* The Reactive SQL Client
  * `io.reactiverse.sqlclient`
  * `io.reactiverse.sdbc`
* The Reactive DB Client
  * `io.reactiverse.dbclient`
  * `io.reactiverse.rdbc`

## Specific operations

* Provide a generic API in the project package that does not provide specific operations
* Each database has its own package and extends the generic API providing specific operations

Examples

```
public interface DbConnection {
}

public interface PgConnection extends DbConnection {
  // Specific stuff here
}
```

Issues : we cannot really do this unless with use self bound generics ,i.e

```
public interface DbClient<C extends DbConnection> {
  void connect(Handler<AsyncResult<C>> handler);
}

public interface PgClient extends DbClient<PgConnection> {
}
```

unless we don't abstract PgClient ?

**Comment by vietj** [comment]
give your opinion here, I'll update the issue description with your ideas

**Comment by mp911de** [comment]
Did you consider using R2DBC as general standard SPI? This way, you get a proper API for all reactive database drivers and you can expose your mature drivers through a Reactive Streams API. With using R2DBC as a first-class citizen you would open up the eco-system for various clients that are chosen by the developer and not mandated by the driver.

**Comment by vietj** [comment]
@mp911de there are no contributions so far to implement the R2DBC specification, feel free to open an issue to make your case, I'll add an "help wanted" tag to it for contributors.

**Comment by mp911de** [comment]
Happy to do so if it makes generally sense from your perspective.

**Comment by dazraf** [comment]
What's the status between R2DBC and ADBA? Last I read: https://www.infoq.com/news/2018/10/springone-r2dbc

**Comment by dazraf** [comment]
Assuming there aren't any nascent standards forming in this space, my preference is for Proposal 1.

**Comment by mp911de** [comment]
ADBA is async and based on CompletableFuture. There’s no release available, only a couple commits for GPL licensed code. R2DBC is based on ReactiveStreams, Apache-licensed and there are milestone releases available as JAR along with a written specification.

 Both specs are trying to solve similar problems and emphasize different properties of how they approach the solution.

**Comment by rgielen** [comment]
IMO exposing a generic API specific to reactive-pg(?)-client and implementing R2DBC are both valid concerns. However, they seem to be orthogonal in the sense that they have independent use cases and can stand side by side.

Given that reactive-pg-client already has an elaborate API for direct usage and given that supporting more backends is around the corner, @vietj's proposal for API generification makes perfect sense. For now I am still undecided which given proposal to prefer, though.

I also strongly support the proposal to implement the R2DBC SPI by @mp911de. R2DBC adoption is on the rise, and it plays very nicely with the reactive Spring Data and Spring Reactive MVC ecosystem, making it very accessible for a huge number of application developers. ADBA on the other hand doesn't seem to get real world adoption any time soon, and I share Mark's opinion that it is currently not favourable to implement a ADBA SPI, both from a technical and licensing standpoint. My personal guess is that either R2DBC becomes what Hibernate was to JPA (read: factual template) or both continue to be driven independently. Given the latter, I predict application developers to vote with their feet, favouring R2DBC broadly unless being tied to Jakarta EE which _might_ integrate ADBA in future versions.

**Comment by vietj** [comment]
@rgielen there is a separate issue about R2DBC, also you can make extra proposal for the API of the PG client, it's an open discussion . Keep in mind that R2DBC (or even this project) won't implement some protocols such as Oracle, so ADBA will remain unavoidable for such databases.

**Comment by mp911de** [comment]
> Keep in mind that R2DBC (or even this project) won't implement some protocols such as Oracle, so ADBA will remain unavoidable for such databases.

I don't think that this would be even remotely necessary. With [Project Loom's](https://wiki.openjdk.java.net/display/loom/Main) fibers there is no need to provide native JDBC drivers for exotic or closed-source database protocols. Putting JDBC on a `Thread` is one story, putting it on a `Fiber` an entirely different one.

**Comment by vietj** [comment]
@mp911de I don't see the relationship with project Loom, can you elaborate ?

**Comment by mp911de** [comment]
The reason why we all build database drivers with non-blocking I/O is that today's database drivers are all blocking (JDBC) and we have use-cases that gain efficiency/scalability by eliminating blocking behavior. Today, we cannot turn blocking APIs into non-blocking ones without creating a new library.

All non-blocking database drivers come up with vendor-specific API as there's no real standard API for non-blocking database access. This makes it difficult (impossible?) for client library authors (O/R mappers, SQL-generating and executing libraries) to come up with convenient libraries so application developers can choose a library they like and want to use in a non-blocking application. That's where R2DBC comes into play. 

While ADBA is not yet a thing, we need a proper API, at least from my perspective, to enable client library authors to build tools that support application developers in their productivity. An API that is powerful, yet simple to be implemented natively or by wrapping an existing database driver – such as PgClient.

I expect Loom to happen eventually and from that time on, we get a thing that can turn blocking APIs into non-blocking ones without sacrificing threads. With a `FiberPool` and the R2DBC API, we can provide an implementation (such as `r2dbc-over-jdbc-on-fibers`) that adapts existing JDBC drivers for databases who do not provide a (non-blocking|reactive|asynchronous|you name it) driver. This way, we get the best of two worlds: Mature JDBC drivers and a streaming-capable API using with end to end non-blocking behavior.

From my perspective, ADBA and [`2dbc-over-adba`](https://github.com/r2dbc/r2dbc-over-adba) are only milestones on the way towards non-blocking database access.

**Comment by vietj** [comment]
here is an attempt to abstract into a generic api with some specific parts https://github.com/reactiverse/reactive-pg-client/tree/api-test-1

- Java generics are not used, it seems it would be too cumbersome here
- Row is not abstracted (and I think it's fine)
- some types have been left as is for now (`PgPreparedQuery`, `PgRowSet`, etc...) and not converted as such types don't use the client/pool/connection types so it was not relevant to rename them for this test

**Comment by vietj** [comment]
I've pushed another commit that move more interfaces to the `sqlclient` package and now it looks ok to me.

- the `io.reactiverse.sqlclient` does not leak into `io.reactiverse.pgclient` for the main types
- the user will need to cast `SqlConnection` to `PgConnection` to access specific API, etc...

I believe the PG  specific datatypes will be removed from `Row` / `Tuple` and can be used using type tokens:

```
<T> T get(int index, Class<T> type);
```

which can used this way:

```
Numeric numeric = row.get(0, Numeric.class);
```

**Comment by vietj** [comment]
I pushed the specific data type decoupling with PG in the branch https://github.com/reactiverse/reactive-pg-client/tree/api-test-1

I think this branch could be merged in the mysql-support branch as I think the result is satisfying for the moment

**Comment by vietj** [comment]
ping @pmlopes @tsegismont @FroMage may you have a look at https://github.com/reactiverse/reactive-pg-client/tree/api-test-1 ?

**Comment by vietj** [comment]
actually squashed and merge in mysql_support branch

**Comment by pmlopes** [comment]
I agree with proposal 1 since it seems to make it easier to add other backends and the API stays simple for the user.

**Comment by vietj** [comment]
@pmlopes can you have a look at the current API in the branch ?

**Comment by BillyYccc** [comment]
Hi, here is a proposal for SQL result handling from my point of view.

Currently type related methods are scattered in `Row` and `Tuple` class, the types include common SQL types and PostgreSQL specific types.
I would like to make the following changes:
* Move type related methods out of `Tuple` class, make `Tuple` a generic class which can handle all kinds of class type.
* `Tuple` should also be immutable, at least mutable methods should not be exposed as public APIs.
* PostgreSQL provides rich type support, so methods of PG specific types should be moved into a new class `PgRow` which extends `Row`, such as array,geometric, only common type related methods are kept in `Row` class, but users may need to cast `Row` to `PgRow` when using PG specific API after changing.
* The current APIs in `Row` are huge and can be simplified a lot. methods like `getBoolean` or `getInteger` can be replaced with `<T> T get(Class<T> type, int pos);` , though methods of array type can not be changed due to the limits of `codegen` support.

This may bring many changes so we need more discussions here.

**Comment by vietj** [comment]
The current Row class has already been changed to accommodate specific types:

- `Tuple` is already now generic and not related to specific PostgreSQL types anymore
- I don't get why it should be immutable since the user can add values to a tuple when building it
- we don't need specific row implementation, instead we need to use the `Class` objects instead, i.e we do `tuple.get(Json.class, 0)`
- commonly used types should keep specific accessors (i.e `getOffsetTime("time")`) there is already such accessors in the current branch

**Comment by BillyYccc** [comment]
Mostly `Tuple` is immutable but this makes sense if being mutable is pratical. I wonder if we should prune the specific accessors as many as possible since the `Row` API are bloated with `getXXX(int index)`, `getXXX(String columnName)` and `addXXX()` methods.

**Comment by vietj** [comment]
I think we should not as long  as it is commonly used type from the JDK API.

they can be also useful when the user uses auto completion, for instance he types "dates" and the date stuff appears in the completion

**Comment by vietj** [comment]
<img width="528" alt="Screen Shot 2019-03-25 at 14 17 55" src="https://user-images.githubusercontent.com/225674/54922680-f1687c00-4f08-11e9-8904-80ff37eb70ed.png">

**Comment by BillyYccc** [comment]
Specific accessors may be useful for `getXXX()` methods, but the implementations for `addXXX()` are the same, why not just use `addValue` for all add methods?
https://github.com/reactiverse/reactive-pg-client/blob/d1a534ab99770aa72d254aa6e3446529d6a4618c/src/main/java/io/reactiverse/sqlclient/impl/ArrayTuple.java#L352-L538

**Comment by vietj** [comment]
yes I wanted to add it, although there is already "add" that is usable.

**Comment by vietj** [comment]
note that `add` is inherited from the `List` interface
